File: ios/Aa2Sdk.m
// Aa2Sdk.m

// import RCTBridgeModule
#if __has_include(<React/RCTBridgeModule.h>)
#import <React/RCTBridgeModule.h>
#elif __has_include(“RCTBridgeModule.h”)
#import “RCTBridgeModule.h”
#else
#import “React/RCTBridgeModule.h” // Required when used as a Pod in a Swift project
#endif

// import RCTEventEmitter
#if __has_include(<React/RCTEventEmitter.h>)
#import <React/RCTEventEmitter.h>
#elif __has_include(“RCTEventEmitter.h”)
#import “RCTEventEmitter.h”
#else
#import “React/RCTEventEmitter.h” // Required when used as a Pod in a Swift project
#endif

@interface RCT_EXTERN_MODULE(Aa2Sdk, NSObject)
RCT_EXTERN_METHOD(sendCMD: (NSString *) cmd)

RCT_EXTERN_METHOD(initAASdk)

RCT_EXTERN_METHOD(disconnectSdk)

+ (BOOL)requiresMainQueueSetup
{
  return NO;
}
@end


@interface RCT_EXTERN_MODULE(Emitter, RCTEventEmitter)

RCT_EXTERN_METHOD(supportedEvents)

@end

File: ios/Aa2Sdk.swift
//
//  Aa2Sdk.swift
//  react-native-aa2-sdk
//
//  Created by Cristian Lungu on 30.07.21.
//

import Foundation
import AusweisApp2

@objc(Emitter)
class Emitter: RCTEventEmitter {
    public static var shared:Emitter?
    
    override init() {
        super.init()
        Emitter.shared = self
    }
    override func supportedEvents() -> [String]! {
        return ["onMessage", "onError", "onSdkInit", "onSdkDisconnect", "onSessionIdReceive", "onCommandSentSuccessfully", "onNewIntentSuccess"]
    }
    
    @objc static override func requiresMainQueueSetup() -> Bool {
        return false
    }
}

struct OkResponse: Encodable {
    var message: String
}

struct ErrorResponse: Encodable {
    var error: String
}


func cb(msg: Optional<UnsafePointer<Int8>>) -> Void {
    if let unwrapped = msg {
        let response = OkResponse(message: String(cString: unwrapped))
        // TODO Better handling here
        let jsonString = try! JSONEncoder().encode(response)
        
        Emitter.shared?.sendEvent(withName: "onMessage", body: String(data: jsonString, encoding: .utf8))
    } else {
        Emitter.shared?.sendEvent(withName: "onSdkInit", body: nil)
    }
}



@objc(Aa2Sdk)
class Aa2Sdk: NSObject {
    func sendNotInitializedErrorEvent() {
        Emitter.shared?.sendEvent(
            withName: "onError",
            body: String(data: try! JSONEncoder().encode(ErrorResponse(error: "SdkNotInitializedException")), encoding: .utf8)
        )
    }

    @objc func sendCMD(_ command: String) {
        if(ausweisapp2_is_running()) {
            ausweisapp2_send(command)
            Emitter.shared?.sendEvent(withName: "onCommandSentSuccessfully", body: nil)
        } else {
            sendNotInitializedErrorEvent()
        }
      
    }
    
    
    @objc func initAASdk() {
        if (!ausweisapp2_init(cb, nil)) {
            Emitter.shared?.sendEvent(
                withName: "onError",
                body: String(data: try! JSONEncoder().encode(ErrorResponse(error: "SdkInitializationException")), encoding: .utf8)
            )
        }
    }
    
    @objc func disconnectSdk() {
        if (!ausweisapp2_is_running()) {
            sendNotInitializedErrorEvent()
        } else {
            ausweisapp2_shutdown()
            Emitter.shared?.sendEvent(withName: "onSdkDisconnect", body: nil)
        }
    }
}
